ts/query.go

```diff
@@ -0,0 +1,411 @@
+		return util.Errorf("data added to dataSpan with mismatched sample duration period.")
+	}
+
+	// Calculate an offset adjustment which normalizes the supplied data into
+	// the dataSpan's time period.
+	adjustment := (data.StartTimestampNanos - ds.startNanos) / ds.sampleNanos
+	rd := normalizedData{data, int32(adjustment)}
+
+	// If all samples in the supplied data set are before normalized offset 0,
+	// do not include it.
+	if rd.offsetAt(len(data.Samples)-1) < 0 {
+		return nil
+	}
+
+	// If the supplied data does not occur after all previously added data,
+	// return an error.
+	if len(ds.datas) > 0 {
+		last := ds.datas[len(ds.datas)-1]
+		if rd.offsetAt(0) <= last.offsetAt(len(last.Samples)-1) {
+			return util.Error("data must be added to dataSpan in chronological order.")
+		}
+	}
+
+	ds.datas = append(ds.datas, rd)
+	return nil
+}
+
+// timestampForOffset returns an appropriate timestamp for the given offset
+// within a dataSpan. Because each offset represents a duration of time and not
+// an exact time, the returned timestamp will fall exactly in the middle of the
+// time slot represented by the offset.
+func (ds *dataSpan) timestampForOffset(offset int32) int64 {
+	return ds.startNanos + (int64(offset) * ds.sampleNanos) + (ds.sampleNanos / 2)
+}
+
+// normalizedData is used to "normalize" an InternalTimeSeriesData object for
+// use in a dataSpan.  This is accomplished by computing a constant offset
+// adjustment which adjusts each Sample offset to be relative to the start time
+// of the dataSpan, rather than the start time of the InternalTimeSeriesData.
+type normalizedData struct {
+	*proto.InternalTimeSeriesData
+	offsetAdjustment int32
+}
+
+// offsetAt returns the normalized offset for the Sample at the supplied index
+// in the InternalTimeSeriesData's Samples collection.
+func (rdc *normalizedData) offsetAt(idx int) int32 {
+	return rdc.Samples[idx].Offset + rdc.offsetAdjustment
+}
+
+// dataSpanIterator is used to iterate through the Samples in a dataSpan. The
+// Samples are spread across multiple InternalTimeSeriesData objects; this
+// iterator thus maintains a two-level index to point to a unique sample.
+type dataSpanIterator struct {
+	*dataSpan
+	offset    int32 // The normalized offset of the current sample within the dataSpan
+	dataIdx   int   // Index of InternalTimeSeriesData which contains current Sample
+	sampleIdx int   // Index of current Sample within InternalTimeSeriesData
+	valid     bool  // True if this iterator points to a valid Sample
+}
+
+// sample returns the InternalTimeSeriesSample value currenly pointed to by this
+// iterator.
+func (dsi *dataSpanIterator) sample() *proto.InternalTimeSeriesSample {
+	if !dsi.valid {
+		return nil
+	}
+	return dsi.datas[dsi.dataIdx].Samples[dsi.sampleIdx]
+}
+
+// advance moves the iterator to point to the next Sample.
+func (dsi *dataSpanIterator) advance() {
+	if !dsi.valid {
+		return
+	}
+	data := dsi.datas[dsi.dataIdx]
+	switch {
```

Comment:
I'm uncomfortable with a switch statement here. I think an if / else if / else would be clearer.